dkc/core/rest/authorized_upload.py

```diff
@@ -0,0 +1,106 @@
+from datetime import timedelta
+import logging
+
+from django.conf import settings
+from django.core import signing
+from django.core.mail import send_mail
+from django.shortcuts import get_object_or_404
+from django.template.loader import render_to_string
+from django.utils import timezone
+from drf_yasg.utils import swagger_auto_schema
+from rest_framework import mixins, serializers
+from rest_framework.decorators import action
+from rest_framework.exceptions import PermissionDenied
+from rest_framework.permissions import BasePermission
+from rest_framework.request import Request
+from rest_framework.response import Response
+from rest_framework.views import View
+from rest_framework.viewsets import GenericViewSet
+
+from dkc.core.models import AuthorizedUpload, Folder
+from dkc.core.permissions import Permission
+
+
+class AuthorizedUploadSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = AuthorizedUpload
+        fields = ['id', 'created', 'creator', 'expires', 'folder', 'signature']
+
+
+class CreateAuthorizedUploadSerializer(serializers.Serializer):
+    folder = serializers.IntegerField()
+
+
+class CompleteAuthorizedUploadSerializer(serializers.Serializer):
+    authorization = serializers.CharField()
+
+
+class CanDeleteAuthorization(BasePermission):
+    message = 'Only the creator of an authorized upload may delete it.'
+
+    def has_object_permission(self, request: Request, view: View, obj: AuthorizedUpload) -> bool:
+        return view.action != 'destroy' or request.user.is_superuser or request.user == obj.creator
+
+
+class AuthorizedUploadViewSet(mixins.DestroyModelMixin, GenericViewSet):
+    queryset = AuthorizedUpload.objects.all()
+
+    permission_classes = [CanDeleteAuthorization]
+    serializer_class = AuthorizedUploadSerializer
+
+    @swagger_auto_schema(
+        responses={201: AuthorizedUploadSerializer},
+        request_body=CreateAuthorizedUploadSerializer,
+    )
+    def create(self, request, pk=None):
+        """Authorize an upload to a folder on your behalf."""
+        serializer = CreateAuthorizedUploadSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        folder = get_object_or_404(Folder, pk=serializer.validated_data['folder'])
```

Comment:
I don't think an invalid `folder_id` should trigger a 404 here. I think that would be a 400.

Reply:
Interesting, can you clarify when you think to use 404 vs. 400 for invalid resource IDs? Is it just based on whether they're in the URL path or some other part of the request?

Reply:
This has been resolved by switching the serializer, but I'd still like to hear your thoughts on this.